matching/README.md

```diff
@@ -0,0 +1,991 @@
+  where Y_1, ..., Y_(l-(h+t)) are fresh anonymous variables that do not appear
+  in the substitution S, and C is the clause of row i except S_i is updated
+  with the binding L |-> (l + 1 : o_j, s_j) and R_i is updated to add
+  (l + 1 : o_j, s_j, h, t)
+* specialize(Len(l), i, List p1, ..., p_h; L; q_1, ..., q_t) =
+  [P_i1, ..., P_i(j-1), p_1, ..., p_h, q_1, ..., q_t, P_i(j+1), ..., P_im] -> C
+  if l == h + t
+  where C is the clause of row i except S_i is updated with the binding
+  L |-> (l + 1: o_j, s_j) and R_i is updated to add (l + 1 : o_j, s_j, h, t)
+* specialize(Len(l), i, List p1, ..., p_h; L; q_1, ..., q_t) = None
+  if l < h + t
+
+We did not define specialization on sort and occurrence vectors over list sorts
+previously, so we also must define this now:
+
+* specialize(Len(l), [o_1, ..., o_m]) =
+  [o_1, ..., o_(j-1), 1 : o_j, ..., l : o_j, o_(j+1), ..., o_m]
+  if s_j is a list sort
+* specialize(Len(l), [s_1, ..., s_m]) =
+  [s_1, ..., s_(j-1), s_1, ..., s_l, s_(j+1), ..., s_m]
+  if s_j is a list sort
+  where s_1, ..., s_l are all equal to the element sort of s_j
+
+Now we have to define `decomposeList`:
+
+```
+def decomposeList(Matrix m, Column j): DT
+  val newO = size : o_j
+  val (maxHead, maxTail) = maxListSize(j)
+  val listCases = compiledCases(m, j)
+  for (Len(l), dt) in listCases:
+     replace (Len(l), dt) with (Len(l), listGet(dt, l, o_j, s_j)) in listCases
+  val listDefault = compiledDefault(m, j)
+  listDefault = listGetEnd(listDefault, maxHead, maxTail, o_j, s_j)
+  listDefault = listGet(listDefault, maxHead, o_j, s_j)
+  val f = the function that returns the size of a list
+  return Function(newO, f, [(o_j, s_j)], Int,
+             Switch(newO, Int, listCases, listDefault))
+
+def listGet(DT dt, Int len, Occurrence o, Sort s): DT
+  if len == 0: 
+    return dt
+  val s' = the element sort of s
+  val f = the function that returns the lenth element of a list
+  return listGet(Function(len : o, f, [(o, s)], s', dt), len-1, o, s)
+
+def listGetEnd(DT dt, Int maxHead, Int len, Occurrence o, Sort s): DT
+  if len == 0:
+    return dt
+  val s' = the element sort of s
+  val f = the function that returns the lenth element from the end of a list
+  return listGetEnd(Function(len+maxHead, f, [(o, s)], s', dt), maxHead, len-1, o, s)
+```
+
+Roughly speaking, `decomposeList` behaves similarly to the default Switch case
+except it switches over the runtime length of the list and makes sure to add
+bindings for each list element being matched on to the substitution. Note that
+in the case of the default, we know that the list must be greater in length
+than any list pattern in the column being decomposed, so we only must decompose
+maxHead elements from the front of the list and maxTail elements from the end,
+since we can guarantee any elements in between will never be referenced by any
+submatrix. This is the reason why we treat the default operation as equivalent
+to specializing on `maxHead + maxTail` elements.
```

Comment:
This kind of explanation is nice to have up-front, even if in an abbreviated form.